### PR TITLE
core: don't dispatch onResponseEntity... events for entities that don't have to be subscribed

### LINKED ISSUE #3530: Slot execution failed in 10.2.1
Hello, 

after upgrading to akka-http 10.2.1, we started observing the following stacktraces in logs. They went away once we downgraded to 10.2.0. Unfortunatelly I was not able to extract a reproducer. 

```
ERROR - [0 (Idle)] Slot execution failed. That's probably a bug. Please file a bug at https://github.com/akka/akka-http/issues. Slot is restarted.
java.lang.IllegalStateException: Cannot [onResponseEntitySubscribed] when in state [Idle]
	at akka.http.impl.engine.client.pool.SlotState.illegalState(SlotState.scala:82)
	at akka.http.impl.engine.client.pool.SlotState.onResponseEntitySubscribed(SlotState.scala:64)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$Event$.$anonfun$onResponseEntitySubscribed$1(NewHostConnectionPool.scala:180)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$Event$.$anonfun$event0$1(NewHostConnectionPool.scala:189)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$Slot.runOneTransition$1(NewHostConnectionPool.scala:277)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$Slot.loop$1(NewHostConnectionPool.scala:370)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$Slot.updateState(NewHostConnectionPool.scala:379)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$Slot.updateState(NewHostConnectionPool.scala:267)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$Slot.onResponseEntitySubscribed(NewHostConnectionPool.scala:256)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$SlotConnection.$anonfun$onPush$4(NewHostConnectionPool.scala:495)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$SlotConnection.$anonfun$onPush$4$adapted(NewHostConnectionPool.scala:495)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$SlotConnection.withSlot(NewHostConnectionPool.scala:441)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$SlotConnection.$anonfun$onPush$3(NewHostConnectionPool.scala:495)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$SlotConnection.$anonfun$onPush$3$adapted(NewHostConnectionPool.scala:493)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1.$anonfun$safely$2(NewHostConnectionPool.scala:609)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1.$anonfun$safeCallback$1(NewHostConnectionPool.scala:608)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1.$anonfun$safeCallback$1$adapted(NewHostConnectionPool.scala:608)
	at akka.stream.impl.fusing.GraphInterpreter.runAsyncInput(GraphInterpreter.scala:466)
	at akka.stream.impl.fusing.GraphInterpreterShell$AsyncInput.execute(ActorGraphInterpreter.scala:498)
	at akka.stream.impl.fusing.GraphInterpreterShell.processEvent(ActorGraphInterpreter.scala:600)
	at akka.stream.impl.fusing.ActorGraphInterpreter.akka$stream$impl$fusing$ActorGraphInterpreter$$processEvent(ActorGraphInterpreter.scala:769)
	at akka.stream.impl.fusing.ActorGraphInterpreter$$anonfun$receive$1.applyOrElse(ActorGraphInterpreter.scala:784)
	at akka.actor.Actor.aroundReceive(Actor.scala:537)
	at akka.actor.Actor.aroundReceive$(Actor.scala:535)
	at akka.stream.impl.fusing.ActorGraphInterpreter.aroundReceive(ActorGraphInterpreter.scala:691)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:577)
	at akka.actor.ActorCell.invoke(ActorCell.scala:547)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:270)
	at akka.dispatch.Mailbox.run(Mailbox.scala:231)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:243)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177) 
```

### PR DESCRIPTION
**Pull request by jrudolph**
For example, for HEAD requests.

Fixes #3530

### CODE REVIEW

File: akka-http-core/src/main/scala/akka/http/impl/engine/client/pool/NewHostConnectionPool.scala [hunk 1]
```diff
@@ -427,6 +420,13 @@ private[client] object NewHostConnectionPool {
               currentTimeoutId = -1
             }
         }
+        private def hasNoEntityStream(response: HttpResponse): Boolean =
+          response.entity match {
+            case _: HttpEntity.Strict        => true
+            case HttpEntity.Default(_, _, e) => e == Source.empty
```
Comment (jrudolph) [inline akka-http-core/src/main/scala/akka/http/impl/engine/client/pool/NewHostConnectionPool.scala:426]: The former `requestContext.request.method == HttpMethods.HEAD` has been replace by this clause. This is somewhat delicate business as we depend on equality to work, which it only does if the body stream is not decorated somewhere else.

File: akka-http-core/src/main/scala/akka/http/impl/engine/client/pool/NewHostConnectionPool.scala [hunk 2]
```diff
@@ -481,33 +481,33 @@ private[client] object NewHostConnectionPool {
 
             withSlot(_.debug("Received response")) // FIXME: add abbreviated info
 
-            response.entity match {
-              case _: HttpEntity.Strict => withSlot(_.onResponseReceived(response))
-              case e =>
-                val (newEntity, StreamUtils.StreamControl(entitySubscribed, entityComplete, entityKillSwitch)) =
-                  StreamUtils.transformEntityStream(response.entity, StreamUtils.CaptureMaterializationAndTerminationOp)
+            // if hasNoEntityStream == true, we don't expect onResponseEntity... events in the state machine above
+            if (hasNoEntityStream(response)) withSlot(_.onResponseReceived(response))
```
Comment (jrudolph) [inline akka-http-core/src/main/scala/akka/http/impl/engine/client/pool/NewHostConnectionPool.scala:485]: The fix here is that we now use the same logic (`hasNoEntityStream`) here and above in the state machine to determine whether to send `onResponseEntity...` event or not, so the bug shouldn't happen any more.

File: akka-http-core/src/main/scala/akka/http/scaladsl/model/HttpEntity.scala [hunk 3]
```diff
@@ -398,8 +398,10 @@ object HttpEntity {
     def withContentType(contentType: ContentType): HttpEntity.Default =
       if (contentType == this.contentType) this else copy(contentType = contentType)
 
-    override def withSizeLimit(maxBytes: Long): HttpEntity.Default =
-      copy(data = Limitable.applyForByteStrings(data, SizeLimit(maxBytes, Some(contentLength))))
+    override def withSizeLimit(maxBytes: Long): HttpEntity.Default = {
+      if (data ne Source.empty) copy(data = Limitable.applyForByteStrings(data, SizeLimit(maxBytes, Some(contentLength))))
```
Comment (jrudolph) [inline akka-http-core/src/main/scala/akka/http/scaladsl/model/HttpEntity.scala:402]: Did I say delicate business? E.g. here we have to special case to preserve `Source.empty` if possible.

File: akka-http-core/src/test/scala/akka/http/impl/engine/client/HostConnectionPoolSpec.scala [hunk 4]
```diff
@@ -190,18 +190,45 @@ class HostConnectionPoolSpec extends AkkaSpecWithMaterializer(
 
         conn1.expectRequestToPath("/2")
       }
-      "not get stuck on HEAD requests if payload is never subscribed" inWithShutdown new SetupWithServerProbes(_.withResponseEntitySubscriptionTimeout(Duration.Inf).withMaxConnections(1)) {
-        pushRequest(HttpRequest(method = HttpMethods.HEAD, uri = "/head"))
-        pushRequest(HttpRequest(uri = "/2"))
-        val conn1 = expectNextConnection()
-        conn1.expectRequestToPath("/head")
+      if (clientServerImplementation != PassThrough) {
```
Comment (jrudolph) [inline akka-http-core/src/test/scala/akka/http/impl/engine/client/HostConnectionPoolSpec.scala:193]: Delicate, what? The test infra adds decorators to the entity in PassThrough mode, but no worries this is less realistic than the other tests.

File: akka-http-core/src/main/scala/akka/http/impl/engine/client/pool/NewHostConnectionPool.scala (hunk 1 above)
Comment (jrudolph) [reply akka-http-core/src/main/scala/akka/http/impl/engine/client/pool/NewHostConnectionPool.scala:426]: After all this is an optimization: for HEAD requests we want to be able to make quick progress even if the user doesn't subscribe entities (quickly enough).

In the worst case, that optimization doesn't work any more and we fall back to pre-10.2.1 times.

**Review by jrudolph — COMMENTED** [review]
Some comments for reviewing below.

**Review by ignasi35 — APPROVED** [review]
LGTM. Thanks for all the reviewing notes @jrudolph, they made the changes a lot more comprehensible.

**Review by johanandren — APPROVED** [review]
Tricky business. LGTM

**Pull request merged**